content: source track draft: simplify and clarify level goals (#1097)
slsa-framework/slsa: docs/spec/draft/source-requirements.md

| Approver | The role that approves a particular change to the source.
| Merger | The role that applies a change to the source. This person may be the proposer or a different trusted person, depending on the version control platform.

## Source Control Platform and Version Control System Requirements

The combination of SCP and VCS MUST provide:

> ### **[Immutable reference]**

Author: changing these to hyperlinkable sections `###`

There exists a deterministic way to identify a particular revision.

This is usually a combination of the repository ID and revision ID.
When the revision ID is a digest of the revision, as in git, nothing more is needed.
When the revision ID is a number or otherwise not a digest, then the repository server MUST guarantee that revisions cannot be altered once created.
The SCP MUST guarantee that repository IDs track the complete history of changes that occur to the source while hosted on the platform.

### **[Identity Management]**

There exists an identity management system or some other means of identifying actors.

The SCP will use these identities for recording change history and writing provenance attestations.

> ### **[Change history]**

Author: This is what we'd proposed in the gdoc. 
It might make more sense to talk about sequences of approved revisions and say it's the SCP's job to explain how they got there. See https://github.com/slsa-framework/slsa/issues/1095

There exists a record of the history of changes conducted on this SCP that went into the revision.

> A merged GitHub pull request is an example of a change record.

Author: git technically tracks revisions, so "change records" are only really "tracked" by pull request tooling if we take this definition.

Each change MUST contain:

>     -   The immutable reference to the new revision.

Author: gh prs are _not_ immutable today -- something that needs to be fixed. 
merged PRs are definitely what most customers would consider to be "source provenance attestations" but they are definitely lacking some of these elements.

Author: I'm not sure how much of this would need to be reiterated in an attestation. 
This section covers what a functional SCP "MUST" be able to do. 

I think the interesting piece is whether we require a history of revisions or a history of changes. 
Practically, git stores revisions but PRs appear to track "changes," or at least that's what is reviewed by approvers.

Author: > merged PRs are definitely what most customers would consider to be "source provenance attestations"

☝️ this is true in the sense of what typical verifiers do today. 

Your typical enterprise process goes something like...

Whereas: 
1. The scp supports branch protection rules. We trust those to be implemented correctly.
1. A trusted team controls the rules for a ref. We trust those folks to implement the right rules and not to abuse their admin powers.
1. The pr tool we use embeds a link to the pr that created it into the commit message, most of the time. 
1. Therefore, I can deploy any revisions I get when I clone a repo at the ref, and hopefully if there are problems, I can use the pr url to understand how this commit got there.

The goals of the slsa source track should be to improve the quality of this process. 
We'll do this by recommending rules and the kind of info that should be attested by scps and pr tools.

Reviewer: I think this is obsolete.

    -   The list of parent revisions.
    -   The change description/justification.
    -   The content of the change.
    -   The [Merger](#source-roles) of the change.
    -   The timestamp when the change was received by the SCP

Administrators have the necessary permissions to replace the source in a known repo.
In mechanical terms, this means changing the source for a known repository ID without publishing a change record.
This includes changing files, history, or changing references in git.
When used as an attack, this is called “repo hijacking” (or “repo-jacking”) and is one of the primary threats source provenance attestations protect against.

> If an organization must change the source without publishing a change record, the organization will need to demonstrate that the change was necessary and executed responsibly.

Author: How do we summarize this better.
It's kind of like, "admins have a lot of power, so please do your maintenance in a public way"

| Trusted person | A human who is authorized by the organization to propose and approve changes to the source.
| Trusted robot | Automation with an authentic identity that is authorized by the organization to propose and/or approve changes to the source.
| Untrusted person | A human who has limited access to the project. They MAY be able to read the source. They MAY be able to propose or review changes to the source. They MAY NOT approve changes to the source or perform any privileged actions on the project.
| Proposer | An actor that proposes (or uploads) a particular change to the source.
| Reviewer / Voter / Approver | An actor that reviews (or votes on) a particular change to the source.
| Merger | An actor that applies a change to the source. This actor may be the proposer.

## Safe Expunging Process

Administrators have the ability to expunge (remove) content from a repository and its change history without leaving a record of the removed content.

> This includes changing files, history, or changing references in git and is used to accommodate legal or privacy compliance requirements.

Reviewer 2: I think one requirement should be that Administrators must use a multi-party authorization process for this. It doesn't solve the transparency side, but at least internally it puts the necessary requirement that this cannot be done unilaterally. E.g. if this end up being a insider repo hijacking, the organization owner would be able to show audit that the multiparty process was followed.

Author: I tend to agree -- the hope was that small teams of 1 could participate in slsa somehow. 
multi-party review would exclude them as a class?

maybe we could differentiate between "any" expunging process and "safe?"

Reviewer: yes, having some process that's visible to repo readers sounds good?  E.g. "an issue that notes that it happened and some description of why (even if it's not possible to provide details)"

Author: github docs have this to say about removing data: https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository

I think the key points are: 
* any form of history re-write counts, including normal flows like force pushing. 
* `git push --force` is totally fine and a good idea for user-owned refspecs, but thou SHALT NOT force to main without following a bunch of process. 
* Rewriting history and skipping "the expungement process" is morally equivalent to shipping an entirely new repository.

Author: * the process needs to include a way to connect the new release / revision to the previous. It probably should include the previous revision id in an artifact somewhere. The refupdate will not contain it iirc.

Author: I took a crack at this below with the "consumable branch" stuff.

When used as an attack, this is called “repo hijacking” (or “repo-jacking”) and is one of the primary threats source provenance attestations protect against.

TODO: Determine how organizations can provide transparency around this process.

## Levels

> ### Level 1: Immutable revisions and Change Management Processeses

Reviewer: ```suggestion
### Level 1: Version controlled
```

I think 'Version controlled' is both 1) is still a fine description for this section, 2) is more human readable (less technical jargon).

Summary: The source is stored and managed through a modern version control system and modified via a codified revision process.

Intended for: Organizations wanting to easily and quickly gain some benefits of SLSA and better integrate with the SLSA ecosystem without changing their source workflows.

Benefits:
Version control solves software development challenges ranging from change attribution to effective collaboration.
It is a software development best practice with more benefits than we can discuss here.

Requirements: The combination of SCP and VCS MUST provide:

#### Revisions are immutable and uniquely identifiable

There exists a deterministic way to identify a particular revision.

This is usually a combination of the repository ID and revision ID.

> When the revision ID is a digest of the contend of the revision (as in git) nothing more is needed.

Reviewer 2: nit: s/contend/content

When the revision ID is a number or otherwise not a digest, then the SCP MUST document how the immutability of the revision is established.
See also [Use cases for non-cryptographic, immutable, digests](https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md#use-cases-for-non-cryptographic-immutable-digests).

#### Identity Management

There exists an identity management system or some other means of identifying actors.
This system may be a federated authentication system (AAD, Google, Okta, GitHub, etc) or custom implementation (gittuf, gpg-signatures on commits, etc).
The SCP MUST document how actors are identified for the purposes of attribution.

Activities conducted on the SCP SHOULD be attributed to authenticated identities.

#### Trusted revision process

There exists a trusted mechanism for modifying the source pointed-to by a [branch](#definitions).

> It MUST NOT be possible to modify the content of a branch without following the documented process.

Reviewer: What are you envisioning for "documented process"?

It could be something like "each project needs to write down in a README how they make changes to their codebase and then have some way to enforce that in the SCP" but that doesn't seem right.  It would be putting too much on the individual project owners. (and it would be really hard to check automatically).  So maybe it's just for the SCP.

Maybe it's more like "The SCP must document how changes can be made to the underlying VCS and what restrictions project owners can place on that process.  When owners configure these restrictions it must not be possible to bypass the restrictions.  E.g. if a project branch is configured to require a pull request prior to merge it must not be possible to modify that branch without going through a pull request."

Author: > What are you envisioning for "documented process"?

For soc 2 style audits, there's a real document somewhere that matters, but I don't think we need that here. I _generally_ think these kind of semantics can just be encoded into the rules and that's good enough. 

Effectively, a git repo is going to be full of a lot of junk. For consumers to make sense of it, an organization needs to explain very basically how they're using git. I think there should be some statement, configuration, or document of process that says or implies:

### My Development Process
1. this repo has some semantically important refspecs
2. "/refs/heads/main" is the one we try to keep healthy. Revisions reachable from main represent a meaningful history of the intent of this repo and have followed our development process. The ref has been minimally protected and we do not allow force pushes to this ref without following a Safe Expunging Process. 
3. (optional) /refs/heads/users/* is where we work and is probably full of random user stuff that is not to be trusted or consumed directly.
4. (optional) /refs/heads/release/* represent revisions held to an even higher scrutiny. 
5. (optional) we :heart: automated tests. our unit testing workflow is called .github/workflows/test.yaml and a clean run is required to merge into main or release branches. Revisions with a green CI are generally better than those without.

An example of a revision process could be: all proposed changes MUST be pre-approved by code experts before being included on a protected branch in git.

The revision process MUST specify the branch a contributor is proposing to update.

> ### Level 2: Provenance

Reviewer: ```suggestion
### Level 2: Verified history
```

I think  there's an argument to be made that 'Provenance' and the other things in this section really just make it possible to verify (and have confidence in the history).  Also since it's less jargony it might be easier for readers?

Author: you're probably right? 
Generally it should be "what are all the things the SCP can attest to that eventually allowed a specific revision to become reachable from this context?"

Author: Provenance attestations should describe how a revision came to be. "Which revisions preceded it?" is critical and is more of what I'd think of as "history," but that's only one of the parts.

> Summary: A consumer can ask the SCP for everything it knows about a specific revision of a repository. The information is provided in a standardized format.

Reviewer: ```suggestion
Summary: The SCP has tamper-proof evidence that documents how revisions in the repository were created and by whom.  This information can be provided to verifiers as needed in standardized formats.
```

The current wording of the summary suggests that any consumer can request this evidence (which will be very detailed).  For OSS that's probably fine but for closed source software it would be a problem as it might leak PII, intellectual property, etc...

I think the real goal is that the SCP and those responsible for checking the SCP have access to this evidence.

Reviewer: But perhaps 'consumer' is really "people and tooling that have access to the source code", which might not include downstream users of the software itself?

Author: I think if you can clone the repo, you should be able to get the attestation (that's how build attestation serving works too). 
There may be cases where the source is packaged into a zip or something, in that case there's lots of wiggle room to inject malicious content and we should fall back to the build track approach: "We don't know what's in there but it was definitely built _here._" etc.

Reviewer: So, "read access to the repo implies read access to the tamper-proof evidence?"  That works for me.

Access to source in a zip should be considered out of scope since SLSA [doesn't consider that to be source](https://slsa.dev/spec/v1.0/terminology#:~:text=A%20ZIP%20file%20containing%20source%20code%20is%20a%20package).

Author: > doesn't consider that to be source

That's an interesting topic on its own -- not all zips are the same I guess. GH lets you download as a zip to avoid the inefficient git clone. It's a common way to consume source in build systems -- I think it's used by actions directly?

It seems like if you have access to the repo revision, you should be able to get the attestations.
I generally think it shouldn't matter how you transferred a repository -- if there is a stable way to recover the repository id, that should be enough to verify. 

If the revision you care about is packaged into some other zip in a way that is not generated / underwritten by the SCP, you'd definitely need to fall back to other package security techniques.

> Intended for: Organizations that need to enforce policy on consumed source.

Reviewer: ```suggestion
Intended for: Organizations that want strong guarantees and auditability of their change management processes.
```

I don't think 'Organizations that need to enforce policy on consumed source.' is good positioning, especially for OSS.  We should probably provide motivation from the frame of reference of the folks that are maintaining the code (they're the ones doing the work!) and less from the side of the consumers.

Benefits: Provides reliable information to policy enforcement tools.

Requirements:

> #### Source provenance attestation

Reviewer: ```suggestion
#### Source attestations
```

We might want to just talk about this in plural form for now since there could be multiple attestations for any revision?

1. The detailed source provenance that says what it was based on, who proposed it, who merged it.
2. The [human review predicate](https://github.com/in-toto/attestation/pull/151) that details how that revision was reviewed
3. The [test results predicate](https://github.com/in-toto/attestation/blob/main/spec/predicates/test-result.md) that details how that revision was tested
4. The #1094 source VSA that just summarizes all of that.

Author: this is a great callout.

Source attestations are associated with the revision identifier delivered to consumers and are a record of everything the SCP knows about the revision's creation process.

> For example, if you perform a `git clone` operation, a consumer MUST be able to fetch the source attestation document using the commit id at the tip of the checked-out branch.

Reviewer: ```suggestion
For example, if you perform a `git clone` operation, a consumer MUST be able to fetch the source attestation documents using the commit id at the tip of the checked-out branch.
```

> Failure of the SCP to return a source attestation for the commit id is the same as saying the revision was not known to have been produced on the SCP.

Reviewer: ```suggestion
Failure of the SCP to return source attestations for the commit id is the same as saying the revision was not known to have been produced on the SCP.
```

#### Trusted revision process requirements

The change management tool MUST provide at a minimum:

##### Strong Authentication

> The strongly authenticated idenenty used to login to the SCP MUST be used for the generation of source provenance attestations.

Reviewer: I think we lost the definition of 'strong authentication' at some point.  (also fixing a misspelling)

Removing "used to login to the SCP" because I think it's a bit over-prescriptive (maybe folks are strongly auth'd via signatures on patches?) and redundant (we're already saying a strongly authed identity must be used to sign the things)?

```suggestion
User accounts that can modify the source or the project’s configuration must use multi-factor authentication or its equivalent.

This strongly authenticated identity MUST be used for the generation of source provenance attestations.
```

See [source roles](#source-roles).

##### Change context

The change management tool MUST record the "target" context for the change proposal and the previous / current revision in that context.

##### Informed Review

The change management tool MUST record the specific code change proposal (a "diff" in git) displayed to reivewers (if any) or instructions to recreate it.
The reviewer is able and encouraged to make an informed decision about what they're approving.
The reviewer MUST be presented with a full, meaningful content diff between the proposed revision and the previously reviewed revision.
It is not sufficient to indicate that a file changed without showing the contents.

##### Verified Timestamps

The change management tool MUST record timestamps for all contributions and review activities.
User-provided values MUST NOT be used.

##### Human-readable change description

The change management tool SHOULD record a description of the proposed change and all discussions / commentary related to it.
All collected content SHOULD be made immutable if the change is accepted.

> ## Choose your own adventure

Reviewer: I _really_ like this section as I think it will let us circumvent some of the problems we've had with requiring code review at high levels.

Author: I agree @TomHennen. I like the content, and it definitely seems correct for it to be out of the source track provenance details.

This week I've been thinking about this. What do you think about trying to move this whole section to the [ossf scorecard project](https://github.com/ossf/scorecard) and cut it from slsa source track entirely?

The scorecard has a lot more opinions about stuff and seems like a logical VSA issuer for some of these configurations that are really good ideas but that don't relate to the source track objective to produce tamper-resistant provenance attestations.

Reviewer: That _is_ an interesting idea.  FWIW I was just chatting with @haydentherapper earlier today about the potential to have Scorecards issue VSAs in support of the source track _as a proof of concept_ before things we're fully fleshed out.  So this might give that effort legs that last longer than just a proof-of-concept.

On a related note I think it might be easier for consumers to define have a single issuing authority who they trust to make assessments of repo revisions.  But I'm not wedded to that and I don't think having Scorecards compute these properties even precludes it.

Author: todo: rename this additional "source controls"

#### Different actors

It MUST NOT be possible for a single actor to control more than one voting accounts.

Should the organization discover that it issued multiple accounts to the same actors, it MUST act to rectify the situation.
For example, it might revoke project privileges for all but one of the accounts and perform retroactive code reviews on any changes where that actors' accounts are the author and/or code reviewer(s).

### Expert Code Reivew

Summary: All changes to the source are pre-approved by experts in those areas.

Intended for: Enterprise repositories and mature open source projects.

Benefits: Prevents mistakes made by developers who are unfamiliar with the area.

Requirements:

#### Code ownership

Each part of the source MUST have a clearly identified set of experts.

#### Approvals from all relevant experts

For each portion of the source modified by a change proposal, pre-approval MUST be granted by a member of the defined expert set.
A approval from an actor that is a member of multiple expert groups may satisfy the requirement for all groups in which they are a member.

### Review Every Single Revision

> Summary: The final revision was reviewed by all relevant experts prior to submission.

Reviewer: Is the idea here that with "two party review" one of the reviewers may have approved a PR while requesting changes.  With 'just' two party review they'd be allowed to submit immediately.  With this requirement those changes would need to be re-approved?

Reviewer: NVM, I see that's explained just below.

Intended for: The highest-of-high-security-posture repos.

Benefits: Provides the maximum chance for experts to spot and reject problems before they ship.

Requirements:

#### Reset votes on all changes

If the proposal is modified after receiving expert approval, all previously granted approvals MUST be revoked.
A new approval MUST be granted from ALL required reviewers.

The new approval MAY be granted by an actor who approved a previous iteration.

### Automated testing

> Summary: The final revision was validated against a suite of vetted automated tests.

Reviewer: ```suggestion
Summary: The final revision was validated against a suite of automated tests.
```

Suggest removing 'vetted' just because I don't think it's clear what it means and it would cause a lot of unnecessary discussion.

Intended for: All organizations and repositories.

Benefits: Automatic testing has many benefits, including improved accuracy, error prevention and reduced workload on your human developers.

> Requirements: For each configured automatic test, results MUST be collected by the change review tool and included in the source provenance attestation.

Reviewer: ```suggestion
Requirements: For each configured automatic test, results MUST be collected by the change review tool, recorded in an attestation, and made available for verification.
```

Suggest we not specify which attestation it get recorded in. There's already [a public predicate type](https://github.com/in-toto/attestation/blob/main/spec/predicates/test-result.md) for recording these results and we should let folks use that if they want?

Policies in this category include things like commit signatures, author / committer formatting restrictions, closed-issue-linkage, etc.

### Repository renames

When a repo is transferred to a new organization ("donated"), or if a repo must be renamed or otherwise have its url changed within the same org, attestations for previous revisions of this repo will no longer be matched because the combination of the repository id and the revision id will have changed.

## Levels

### Level 1: Version controlled

Summary:
The source is stored and managed through a modern version control system.

Intended for: Organizations currently storing source in non-standard ways who want to quickly gain some benefits of SLSA and better integrate with the SLSA ecosystem with minimal impact to their current workflows.

Benefits:
Migrating to the appropriate tools is an important first step on the road to operational maturity.

Requirements:

#### Use modern tools

The organization MUST manage the source using tools specifically designed to manage source code.
Tools like Perforce, GitLab, GitHub, Azure Repos are all great examples.

> Storing revisions as differently-named zips in folders on a laptop does not count.

Reviewer 3: It is sad how often I still see this.

Author: I agree. this comment actually looks a little dismissive / snide to me on second read, but the intention was to clearly differentiate between true version control tools and custom built solutions. 
Is there a better way to phrase this?

Reviewer 4: I agree that local, unauthenticated storage should not fall under the SCP category. But given the definition of SCP above, I wonder if only calling out the "commercial" services inadvertently excludes cases of enterprises or projects (e.g., Linux kernel) having self-hosted infrastructure that does otherwise meet the criteria for SCP (hosted service, authenticated, access controlled, etc.).

Reviewer: I made a suggestion to try to address @marcelamelara's comment

Reviewer: Zac applied that suggestion so I think we're good?  (and can tweak the language in follow up PRs)

Author: great callout: we should probably consider removing those cloud names and just saying git, svn, perforce,  

`Tools like git, Perforce, Subversion are great examples. They may be self-hosted or hosted in the cloud using vendors like GitLab, GitHub, BitBucket, etc.`

Reviewer: I like that.

### Repository renames

When a repo is transferred to a new organization ("donated"), or if a repo must be renamed or otherwise have its url changed within the same org, attestations for previous revisions of this repo will no longer be matched because the combination of the repository id and the revision id will have changed.

## Levels

Many examples in this document use the [git version control system](https://git-scm.com/), but use of git is not a requirement to meet any level on the SLSA source track.

### Level 1: Version controlled

Summary:
The source is stored and managed through a modern version control system.

Intended for: Organizations currently storing source in non-standard ways who want to quickly gain some benefits of SLSA and better integrate with the SLSA ecosystem with minimal impact to their current workflows.

Benefits:
Migrating to the appropriate tools is an important first step on the road to operational maturity.

Requirements:

#### Use modern tools

The organization MUST manage the source using tools specifically designed to manage source code.

> Tools like Perforce, GitLab, GitHub, Azure Repos are all great examples.

Reviewer: ```suggestion
Tools like Perforce, GitLab, GitHub, Azure Repos, and self-hosted git _servers_ are all great examples.
```

> Tools like Perforce, GitLab, GitHub, Azure Repos, and self-hosted git _servers_ are all great examples.

Author: ```suggestion
Tools like git, Perforce, Subversion are great examples. 
They may be self-hosted or hosted in the cloud using vendors like GitLab, GitHub, Bitbucket, etc.
```

> Storing revisions as differently-named zips in folders on a laptop does not count.

Author: maybe something like this? @marcelamelara @TomHennen ?

```suggestion
When self-hosting a solution, local, unauthenticated storage is not acceptable.
```

Summary: The final revision was reviewed by all relevant experts prior to submission.

Intended for: The highest-of-high-security-posture repos.

Benefits: Provides the maximum chance for experts to spot and reject problems before they ship.

Requirements:

#### Reset votes on all changes

If the proposal is modified after receiving expert approval, all previously granted approvals MUST be revoked.
A new approval MUST be granted from ALL required reviewers.

The new approval MAY be granted by an actor who approved a previous iteration.

### Automated testing

Summary: The final revision was validated against a suite of automated tests.

Intended for: All organizations and repositories.

Benefits: Automatic testing has many benefits, including improved accuracy, error prevention and reduced workload on your human developers.

Requirements: For each configured automatic test, results MUST be collected by the change review tool, recorded in an attestation, and made available for verification.

For example, you may configure a "required GitHub Actions workflow" to run your test suites.
Only change proposals with a successful workflow run id would be allowed to be submitted.

> ### Only merge the diff that was reviewed

Author: this is the kind of content (once polished up) that seems valuable to smaller shops -- I'd love for it to exist somewhere eventually.

| Trusted person | A human who is authorized by the organization to propose and approve changes to the source.
| Trusted robot | Automation with an authentic identity that is authorized by the organization to propose and/or approve changes to the source.
| Untrusted person | A human who has limited access to the project. They MAY be able to read the source. They MAY be able to propose or review changes to the source. They MAY NOT approve changes to the source or perform any privileged actions on the project.
| Proposer | An actor that proposes (or uploads) a particular change to the source.
| Reviewer / Voter / Approver | An actor that reviews (or votes on) a particular change to the source.
| Merger | An actor that applies a change to the source. This actor may be the proposer.

> ## Safe Expunging Process

Author: A priority for this section is to document that this process is allowed. 
different teams and tech stacks may have different approaches to the problem. 

A DMCA takedown request will probably be addressed by following an agreedupon process. That process should be documented itself and followed such that auditors are satisfied? unclear.

Administrators have the ability to expunge (remove) content from a repository and its change history without leaving a record of the removed content.
This includes changing files, history, or changing references in git and is used to accommodate legal or privacy compliance requirements.
When used as an attack, this is called “repo hijacking” (or “repo-jacking”) and is one of the primary threats source provenance attestations protect against.

On the git VCS, force pushes allow you to remove data from a branch.
If a branch has been identified as consumable branch, force pushes to that branch must follow the safe expunging process.

TODO: Determine how organizations can provide transparency around this process.
At a minimum the organization would need to declare why data was removed from the branch.

> ## Levels

Author: thoughts: 

1. there are still customers who do not use version control whatsoever
1. a specific source system can be summed into a revision id somehow
1. a revision should have a canonical location
1. that location + revision id can be associated with other data including attestations and slsa levels
2. zac: most teams would start with zips for these revisions? tom: would not be happy with that. 

#### level 1: uses the correct tool
1. The idea should be that you have taken the first concrete step in the right direction. 
1. source zips in onedrive do not count!
2. Any tool built specifically to manage source should count. 

#### level 2: using the tool to protect the history of specific branches
1. starting to use the tool correctly
2. authentication starts to matter
3. basic default branch protections (block force push and deletes, block repo deletion / url renames, safe expunging process). Set up a security posture for a specific context. 

#### level 3: writing attestations
1. tamper resistance
2. strong authn
3. verified timestamps
5. change context (which branches were updated)

#### Canonical location

The source MUST have a location where the "official" revisions are stored and managed.

#### Revisions are immutable and uniquely identifiable

This requirement ensures that a consumer can determine that the source revision they have is the same as a canonical revision.
The combination of SCP and VCS MUST provide a deterministic way to identify a particular revision.

Virtually all modern tools provide this guarantee via a combination of the repository ID and revision ID.

##### Repository IDs

The repository ID is generated by the SCP and MUST be unique in the context of that instance of the SCP.

##### Revision IDs

When the revision ID is a digest of the content of the revision (as in git) nothing more is needed.
When the revision ID is a number or otherwise not a digest, then the SCP and VCS MUST document how the immutability of the revision is established.
The same revision ID MAY be present in multiple repositories.
See also [Use cases for non-cryptographic, immutable, digests](https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md#use-cases-for-non-cryptographic-immutable-digests).

> ### Level 2: Branch History

Author: good question.
my initial thoughts are that we're trying to get across the following concepts:
* teams can have more than one branch
* teams may need to indicate that consumers can / should / must ignore commits on users/* and only ship commits on /releases/* because branches have different security postures
* only some branches have protected history, IE, we allow force push to user branches. 
* the logical VSA for this rule would need to verify that "the previous revision id is reachable from this new revision id" (IE: there was no potential for data loss due to force push or repo hijack)

I'm happy with either name -- what do other folks think?

> ### Level 2: Branch History

Reviewer: Updated suggestion to 'protected branch history'

> ### Level 2: Branch History

Author: actually, let's punt this to the next round. Tracked here: https://github.com/slsa-framework/slsa/issues/1112

##### Continuity

For all branches intended for consumption, whenever a branch is updated to point to a new revision, that revision MUST document how it related to the previous revision.
Exceptions are allowed via the [safe expunging process](#safe-expunging-process).

On VCS like git, the organization MUST enable branch protections that prohibit updating the branch to point to revisions that are not direct descendants of the current revision.
At a minimum, this typically means preventing "force pushes" and "branch deletion."

It MUST NOT be possible to delete the entire repository (including all branches) and replace it with different source.
Exceptions are allowed via the [safe expunging process](#safe-expunging-process).

#### Identity Management

There exists an identity management system or some other means of identifying actors.
This system may be a federated authentication system (AAD, Google, Okta, GitHub, etc) or custom implementation (gittuf, gpg-signatures on commits, etc).
The SCP MUST document how actors are identified for the purposes of attribution.

Activities conducted on the SCP SHOULD be attributed to authenticated identities.

> ### Level 3: Source Provenance Attestations

Reviewer: optional:

```suggestion
### Level 3: Tamperproof
```

I think 'tamperproof' gets at the point of this level pretty well?

Author: hmm, well, that kind of reminds me that this text doesn't talk about writing an itoto doc or signing it but does say "provided in a standardized format". 

probably should add a section about the spec and how to sign it eventually.

The key idea here imo is that this is the list of claims from the SCP's perspective. 
It's the identities verified by the SCP and a bunch of data collected by the change management tooling and git server. 
After achieving this level, you'll have a reliable platform and language for writing many kinds of security claims about source revisions. 

To @TomHennen meta point, I do kind of like the idea of rolling tamper-proof into this level because there's not much the user can do to move the needle here -- these are requirements placed on GitLab/Hub, Azure repos etc. They need to build all the tooling required to emit these docs and they should probably sign them when they do.

Author: @TomHennen I guess it should be "Tamper-resistant attestations"?

Theoretically the revision id in git is what produces tamper-resistant git histories, so it seems like what we're talking about here is the claims document.

Reviewer: I think it's both the claims document and the verified identity (both help contribute to the tamperproof nature).

The advantage that a signed claim has it that it's signed by some authority.  The git revision id might not be tamperable, but there's no indication about how it was produced.  Having attestations sign the thing help solve that problem.

Author: I think we're in agreement, but there's some overlap.
in this pitch it seems to sort of work out that: level 2 minimal policies for "branch histories" gives the the tamper resistance for the code. level 3 is keeping track of other security-related details about each revision and then providing them to policy engines.

Author: tracked here: https://github.com/slsa-framework/slsa/issues/1112

Summary:
A consumer can ask the SCP for everything it knows about a specific revision of a repository.

> The information is provided in a standardized format.

Reviewer: We don't actually have the formats defined yet so maybe we can relax this some to 'documented'?
```suggestion
The information is provided in a documented & tamperproof format.
```

Intended for:
Organizations that want strong guarantees and auditability of their change management processes.

Benefits:
Provides reliable information to policy enforcement tools.

Requirements:

> #### Source provenance attestations

Reviewer: ```suggestion
#### Source attestations
```

Suggest dropping 'provenance' since that often has a specific meaning in SLSA.  I also think 'source attestations' gets the point across with less jargon?

Source attestations are associated with the revision identifier delivered to consumers.
They are a statement of fact from the perspective of the SCP.
They contain information about how a specific revision was created and how it came to exist in its present context.

For example, if a consumer is authorized to access source on a particular branch, they MUST be able to fetch the source attestation documents for revisions in the history of that branch.

> It is possible that an SCP can make no provenance claims about a particular revision.

Reviewer: ```suggestion
It is possible that an SCP can make no claims about a particular revision.
```

Author: @TomHennen I think the SCP could always make claims about when a commit was received. This is specifically that the SCP doesn't know what process created the revision.

For example, this would happen if the revision was created on another SCP.

> #### Change management process
> 
> The repo must define how the content of a [branch](#definitions) is allowed to change.
> It MUST NOT be possible to modify the content of a branch without following its documented process.
> 
> The change management tool MUST provide at a minimum:

Reviewer: I think 'change management process' sounds like it's much broader than what is actually described here.

How do you feel about this?

```suggestion
#### Reliable metadata

The SCP must provide reliable metadata about the revisions it covers.

The change management tool MUST provide at a minimum:
```

I think the

> The repo must define how the content of a [branch](#definitions) is allowed to change.
> It MUST NOT be possible to modify the content of a branch without following its documented process.

Text is actually covered under L2 now given that the repo has to indicate which branches are protected?

Author: I think this is what would need to be covered: 

* A target branch has some minimal branch protections on it.
* a user will push up a number of unshippable commits on some user-controlled topic branch
* a pull request is opened to review the net delta between the tip of those commits and the tip of the target branch
* tests, checks, reviews, comments, and votes will occur on the proposed merge commit in the context of the PR 
* the SCP must record all security metadata related to the creation of the final commit representing the change managed by this pr. 

I guess the critical question is whether you MUST have a pull request context to achieve this level. The PR is the context where the SCP can consider changes like this and might be willing to sign off on how the final commits were created. 

I think yes, for tools like github pull requests (with or without required review) are necessary to achieve this level. Without them, the PR and SCP tools will not be involved in creation of the final merge commit, and will have a hard time attesting anything about revision creation process.

Author: tracked here: https://github.com/slsa-framework/slsa/issues/1113

Examples include a git commit's "author" and "committer" fields and a gpg signature's "user id."
These forms of identity are user-provided and not typically verified by the source provenance attestation issuer.

See [source roles](#source-roles).

##### Context

The change management tool MUST record the specific code change (a "diff" in git) or instructions to recreate it.
In git, this typically defined to be three revision IDs: the tip of the "topic" branch, the tip of the target branch, and closest shared ancestor between the two (such as determined by `git-merge-base`).

The change management tool MUST record the "target" context for the change proposal and the previous / current revision in that context.
For example, for the git version control system, the change management tool MUST record the branch name that was updated.

Branches may have differing security postures, and a change can be approved for one context while being unapproved for another.

##### Verified Timestamps

The change management tool MUST record timestamps for all contributions and review-related activities.
User-provided timestamps MUST NOT be used.

> ## Choose your own adventure / Typical Source Controls and their Applications

Reviewer 5: NOTE: there is a discussion in Slack that proposed removing these from the specification and instead including content like this in the [scorecard](https://github.com/ossf/scorecard?tab=readme-ov-file#scorecard-checks)

Reviewer: I think it might be helpful to have SLSA Source Track define how these things can be communicated to downstream users (e.g. in the VSA from #1094).

In either case Scorecard can definitely be used to _attest_ to them.

### Review Every Single Revision

Summary: The final revision was reviewed by all relevant experts prior to submission.

Intended for: The highest-of-high-security-posture repos.

Benefits: Provides the maximum chance for experts to spot and reject problems before they ship.

Requirements:

#### Reset votes on all changes

If the proposal is modified after receiving expert approval, all previously granted approvals MUST be revoked.
A new approval MUST be granted from ALL required reviewers.

The new approval MAY be granted by an actor who approved a previous iteration.

### Automated testing

Summary: The final revision was validated against a suite of automated tests.

Intended for: All organizations and repositories.

Benefits: Automatic testing has many benefits, including improved accuracy, error prevention and reduced workload on your human developers.

Requirements: For each configured automatic test, results MUST be collected by the change review tool, recorded in an attestation, and made available for verification.

For example, you may configure a "required GitHub Actions workflow" to run your test suites.

> Only change proposals with a successful workflow run id would be allowed to be submitted.

Reviewer: nit: I don't think this meets the "recorded in an attestation" requirement?

### Revision process

There exists a trusted mechanism for modifying the source pointed-to by a [branch](#definitions).
For each [branch](#definitions), the SCP MUST record and keep the full history of changes conducted on this SCP, with exceptions allowed following the [Safe Expunging Process](#safe-expunging-process).

The revision process MUST:

-   Provide an accurate description of the currently proprosed change, or instructions to recreate it.
-   Provide the ability to review a change before it is accepted.
-   Provide the ability to require pre-approval from specific actors before a change proposal is accepted.
-   Record all actors that contributed to the process, see [Source Roles](#source-roles).
-   Record timestamps of critical activities including process start, process completion, reception of change proposals by the SCP, and reviews.
-   Record the specific state of the process when each approval was granted. This is most relevant when the proposal content is allowed to change after aprovals have been granted.

> ### Additional features

Author: I think this needs to be cut and just added as requirements of higher source levels unfortunately.

| Approver | An actor that approves a particular change to the source.
| Merger | An actor that applies a change to the source. This typically involves creating the new revision and updating a branch. This person may be the proposer or a different trusted person, depending on the version control platform.

## Source Control Platform and Version Control System Requirements

The combination of SCP and VCS MUST provide:

### **[Immutable reference]**

There exists a deterministic way to identify a particular revision.

This is usually a combination of the repository ID and revision ID.
When the revision ID is a digest of the revision, as in git, nothing more is needed.
When the revision ID is a number or otherwise not a digest, then the repository server MUST guarantee that revisions cannot be altered once created.
The SCP MUST guarantee that repository IDs track the complete history of changes that occur to the source while hosted on the platform.

> ### **[Identity Management]**

Author: @arewm, I think this is the (pretty terse) summary of the discussion we had in the slsa meeting today. 

I think "which identity did something" is always a claim from the perspective of the tool that chose the authentication tooling. 
IE, Azure repos may use AAD or github may use its own identity system, but both of their claims will be of the form "according to us and systems by which we enforce authn, these actors performed XYZ."

I'm not sure slsa would need its own authn spec?

Author: As a practical example, the "identities who contributed to a source revision" are not always independently verifiable. The authoritative claim comes from the SCP. 

This is convenient and feels like a nice corollary to the build's OIDC token -- it has a bunch of non-independently-verifiable "claims" and are as trustworthy as the issuer.

Reviewer 6: IMO, we should be a bit careful here because we do have the building blocks for independently verifiable identities--signatures. We could leverage signed commits / pushes (eg: you can authenticate that a push to a repository came from me rather than taking the SCP's authentication's word for it) or signed attestations (eg: code review attestations, perhaps integrated into tooling like gittuf, git-appraise, `gitsign attest`, or `gh pr review` even) one day in a future higher level of this track, but in the meantime perhaps we want to be careful not to paint into a corner?

I like the text you've added:

> There exists an identity management system or some other means of identifying actors.

Perhaps this can be accompanied by examples of both the SCP model and the signatures model.

There exists an identity management system or some other means of identifying actors.

> The SCP will use these identities for recording change history and writing provenance attestations.

Author: we should probably also include that an scp will use them to support identity-based rules (like who can review a pr).

In the scenario where a verifier wants to know who contributed to a revision, it's likely that they'll cross-reference with other data sources, like identities with laptops which were known to be compromised during a specific time period. 

My inclination is to say that the scp would not write that fidelity of info into a regular source provenance attestation. 
We'd reserve per-commit metadata for extreme cases or serve it through some other endpoint for teams that need it.

Reviewer 6: This is more a response to the text rather than your comment above.

> The SCP will use these identities for recording change history and writing provenance attestations.

Do you reckon we need to justify what the identities are used for here? Is it enough to say we need some mechanism for identity management and discuss at different levels how they're used?

I ask because I think this may depend on what source provenance ends up looking like. I agree that the SCP is in a position to generate a single provenance document for some change, but in some critical repositories, it may also be necessary to decompose the change into constituent commits/pushes, reviews, test runs, etc. that are performed by different entities that can each generate signed attestations. So a source provenance may in fact be a bundle that contains 1+ code review attestations signed by reviewers, a test run attestation signed by the CI workflow, and so on. This is related to my point above about whether we assume identity is independently verifiable or not as well as https://github.com/slsa-framework/slsa/issues/1042#issuecomment-2214373236. In on-premises SCP deployments, allowing for this decomposing could in future enable applying "trust but verify" to the SCP and its administrators.

Author: > Do you reckon we need to justify what the identities are used for here? Is it enough to say we need some mechanism for identity management and discuss at different levels how they're used?

interesting point -- maybe not! it is probably important to note that the identities used for rules enforcement will be whatever system is trusted by the change management tooling, which may be different than git metadata. 

in practice, this is the difference between the claimed "author" of a commit (about which mainstream SCPs know nothing) and the definitely-involved "pusher" of those commits. 

I'd also be happy summarizing this elsewhere as something like: 
```
source provenance attestations never claim authorship. They only know which actor authenticated to change a specific repository. 
```

This is important for two reasons. 
1. sometimes commit metadata is wrong. or we squash it away and garbage collect the constituent commits.
2. sometimes you need to move a ref to point to someone else's work. This happens a lot in forks.

Reviewer 6: Agree with all of your points! I'm only wary of placing authn responsibilities in any one party when we may want to assign those responsibilities based on the source level further down the line. I think we could just leave it at "some mechanism for identity management / authentication is required" here, then.

In mechanical terms, this means changing the source for a known repository ID without publishing a change record.
This includes changing files, history, or changing references in git.
When used as an attack, this is called “repo hijacking” (or “repo-jacking”) and is one of the primary threats source provenance attestations protect against.
If an organization must change the source without publishing a change record, the organization will need to demonstrate that the change was necessary and executed responsibly.

### **[Change Management]**

There exists a trusted mechanism for modifying the canonical source through a **revision process**.

The revision process MUST record at least:

-   The identities of the proposer, reviewers (if any), and merger (if different from the proposer).
-   Timestamps of change submission. If a change is reviewed, then the change history MUST also include timestamps for any reviews.
-   The specific change reviewed during the revision process or instructions to recreate it. In git, this might be the two compared object ids and the computed best merge base between them at the time of review.

### Additional features

> The combination of SCP and VCS SHOULD additionally provide:

Author: I'm not sure if this is the right place for this list of "shoulds." 
I think I vote to cut it and move it to a section about verifying source control platforms.

| Approver | An actor that approves a particular change to the source.
| Merger | An actor that applies a change to the source. This typically involves creating the new revision and updating a branch. This person may be the proposer or a different trusted person, depending on the version control platform.

> ## Safe Expunging Process

Author: What's some good prior art for this section? 

In practice, we only need this data for revisions that are shipped, or are being considered for shipping. 
This places a pretty convenient recency bias on the amount of data that MUST be kept. 

On the other hand, sometimes repos don't ship new releases for years -- do we need some kind of aging-out process?

Maybe a repo "must" keep 100 releases worth of detailed data, after which it's best effort or something.

| Approver | An actor that approves a particular change to the source.
| Merger | An actor that applies a change to the source. This typically involves creating the new revision and updating a branch. This person may be the proposer or a different trusted person, depending on the version control platform.

## Safe Expunging Process

(placeholder text)

> Administrators have the necessary permissions to replace the source in a known repo without publishing a change record.

Reviewer: To me this seems to prescribe the 'how' too much and we'd be better off describing the effect that's needed?

```suggestion
Administrators have the ability to expunge (remove) content from a repository and its change history without leaving a record of the removed content.
```

This includes changing files, history, or changing references in git.

When used as an attack, this is called “repo hijacking” (or “repo-jacking”) and is one of the primary threats source provenance attestations protect against.

> If an organization must change the source without publishing a change record, the organization will need to demonstrate that the change was necessary and executed responsibly.

Reviewer: I'd suggest we leave this to the specific levels, and maybe a follow on PR.

```suggestion
FUTURE WORK: Determine how organizations can provide transparency around this process.
```

Reviewer: Works for me.

## Source Control Platform and Version Control System Requirements

The combination of SCP and VCS MUST provide:

### Immutable reference

There exists a deterministic way to identify a particular revision.

This is usually a combination of the repository ID and revision ID.
When the revision ID is a digest of the revision, as in git, nothing more is needed.

> When the revision ID is a number or otherwise not a digest, then the repository server MUST guarantee that revisions cannot be altered once created.

Reviewer: The strength of the guarantee will be a constant outstanding question.  I'd suggest weaker language here and then move strength of the guarantee to the levels?

Alternatively maybe with the weakened language it will be fine?

```suggestion
When the revision ID is a number or otherwise not a digest, then the SCP MUST document how the immutability of the revision is established.  See also [Use cases for non-cryptographic, immutable, digests](https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md#use-cases-for-non-cryptographic-immutable-digests).
```

> The SCP MUST guarantee that repository IDs track the complete history of changes that occur to the source while hosted on the platform.

Reviewer: re 'guarantee' maybe it's worth having a generic section that talks about how these guarantees are made and how sure the SCP must be?

Author: that's a good call out. I would think it falls mostly under "trust the platform" but we might want to point that out specifically.

### Identity Management

There exists an identity management system or some other means of identifying actors.
This system may be a federated authentication system (AAD, Google, Okta, GitHub, etc) or custom (gittuf, gpg-signatures on commits, etc).
SCPs SHOULD pick one and use a single identity management system when issuing content in provenance attestations.

> The SCP will use these identities to:

Reviewer 5: Should we explicitly call out the identity of the SCP itself? Being able to describe how consumers would be able to leverage/interpret information from the SCP identity vs. actor identities seems like it would be beneficial.

Author: Are identities controlled by the repos or pull request services inherently "more" trustworthy than other identities? Maybe!

-   Implement actor-based rules (such as requiring review from code experts).
-   Record contributors to a code revision process (in git, this may differ from what is in the commit metadata).
-   Refer to actors in issued source provenance attestations.

### Revision process

There exists a trusted mechanism for modifying the source pointed-to by a [branch](#definitions).
For each [branch](#definitions), the SCP MUST record and keep the full history of changes conducted on this SCP, with exceptions allowed following the [Safe Expunging Process](#safe-expunging-process).

> The revision process MUST:

Reviewer: Maybe some of these can be made level specific?

Author: I propose that this set is the minimum that is meaningfully better than no revision process. Is there a specific point that is too strong.

Reviewer: 'full history'?

I could see only retaining a fully history at higher levels and having something shorter (e.g. 18 months) at lower levels?

Ability to review a change, and pre-approval from specific actors seem too strict for Level 1 (though people should probably consider those things when making choices)

Author: I don't disagree with that 😂, but would there be a middle ground product like that? 
EG, if we made that defintion of peer review "level 2" would level one just be "was reviewed by an _**authenticated**_ actor", potentially after the fact, with no other qualifications?

Reviewer: So there's a long history regarding requirements for peer review and SLSA.  I think the only reasonable paths forward are either:

1) requiring peer review at the highest level only
2) requiring peer review at the highest level only, unless you're an OSS project and then <something else>

I think the original formulation for the source track also only included the review process at the highest level so we're not so far off?

-   Provide an accurate description of the currently proprosed change, or instructions to recreate it.
-   Provide the ability to review a change before it is accepted.
-   Provide the ability to require pre-approval from specific actors before a change proposal is accepted.
-   Record all actors that contributed to the process, see [Source Roles](#source-roles).
-   Record timestamps of critical activities including process start, process completion, reception of change proposals by the SCP, and reviews.
-   Record the specific state of the process when each approval was granted. This is most relevant when the proposal content is allowed to change after aprovals have been granted.

### Additional features

The combination of SCP and VCS SHOULD provide:

-   A mechanism for assigning roles and/or permissions to [actors](#source-roles).

> -   A mechanism for including code review in the [revision process](#revision-process).

Reviewer: These seems like a dupe of "-   Provide the ability to review a change before it is accepted."

-   Two-factor authentication for the [identity management system](#identity-management).
-   Audit logs for sensitive actions, such as modifying security controls.

> -   A mechanism to define code ownership for all files in the source.

Reviewer: Ownership seems like something we should define somewhere or expand on it here?

| Approver | An actor that approves a particular change to the source.
| Merger | An actor that applies a change to the source. This typically involves creating the new revision and updating a branch. This person may be the proposer or a different trusted person, depending on the version control platform.

## Safe Expunging Process

Administrators have the ability to expunge (remove) content from a repository and its change history without leaving a record of the removed content.
This includes changing files, history, or changing references in git.
When used as an attack, this is called “repo hijacking” (or “repo-jacking”) and is one of the primary threats source provenance attestations protect against.

TODO: Determine how organizations can provide transparency around this process.

## Levels

### Level 1: Version controlled

> Summary: The source is stored and managed through a modern version control system with a revision process.

Author: I agree -- for this PR, I just moved the "Responsibly Version Controlled" section to the definition of L1. 

You could squint and say that a merged pull request is a kind of source provenance statement -- it's the only kind we have today without new feature work. 
It's definitely "evidence" that could be consumed by a VSA writer.

Intended for: Organizations that are unwilling or unable to host their source on a source control platform. If possible, skip to Level 2.

Benefits: Version control solves software development challenges ranging from change attribution to effective collaboration.

> It is a software development best practice with more benefits than we can discuss here.

Author: I agree -- it's definitely true that with just merged prs (or other artifacts of completed change management processes) you can get a long way. 
We'll need to decide if provenance attestations are referenced in level 1 and if "merged prs" can count as "mutable provenance attestations" or not.

> It is a software development best practice with more benefits than we can discuss here.

Reviewer 6: Oh I agree with that. In the gittuf case I link, we explicitly create an in-toto statement that is then signed with an online key. We're using this to bridge SCP workflows such as developers merging using github with metadata that is also verifiable using gittuf policies. Another one that's currently an open PR is an attestation created by a gittuf github app that records approvers on a pull request as a signed in-toto attestation (signed by the app which must be trusted of course), and this metadata can again be used to meet review thresholds set in gittuf policies.

> It is a software development best practice with more benefits than we can discuss here.

Author: > I think it's totally fine to consider "merged PRs" as acceptable 'evidence' or 'metadata' (especially at Level 1) but I think it would make things more clear if we didn't call things that weren't signed statements as 'Provenance' or 'Attestations'.

☝️ I think this really shaping up to be the primary difference between the two levels. 
level "1" is the wishy-washy stuff you can get off the shelf today, and level 2 "attestations" will really require some support from bigger SCPs to generate, sign and setup the apis for serving attestations based on revision id. 

Prs / branch protection rule results today are not signed, not in a standardized format, and (at least on gh) mutable. To me this means they might be some kind of evidence, but they do not meet the bar for "attestations" in the same sense as our build attestations.

The SCP MUST guarantee that repository IDs track the complete history of changes that occur to the source while hosted on the platform.

#### Identity Management

There exists an identity management system or some other means of identifying actors.
This system may be a federated authentication system (AAD, Google, Okta, GitHub, etc) or custom (gittuf, gpg-signatures on commits, etc).
SCPs SHOULD pick one and use a single identity management system when issuing source provenance attestations.

When there are conflicting identity claims the authenticated identity MUST be used.
For example in a single git commit the "author", "committer," and the gpg signature's "user id" may be different, and they may all be different than the authenticated identity used to push the commit to the SCP.

#### Revision process

There exists a trusted mechanism for modifying the source pointed-to by a [branch](#definitions).

> For each [branch](#definitions), the SCP MUST record and keep the full history of changes conducted on this SCP, with exceptions allowed following the [Safe Expunging Process](#safe-expunging-process).

Author: we could soften it to "must have a defined change process of any kind", but it seems ok to me to say that "keeping the history of those past revisions" is pretty close to the definition of version control. 

Is that what you were thinking?

> For each [branch](#definitions), the SCP MUST record and keep the full history of changes conducted on this SCP, with exceptions allowed following the [Safe Expunging Process](#safe-expunging-process).

Reviewer: Yeah we talked about this the other day and I think the resolution was basically "by default all SCPs/VCSs keep the entire revision history and in fact getting rid of some of that history after a short amount of time would be pretty difficult to do safely, so we're better off just requiring it from the start and we can see if anyone complains"

But now I'm also remembering some discussion we had where you said something like "some users want to be able to get rid of 'stuff' after a certain amount of time to save storage costs".  I don't remember what 'stuff' was though?

The revision process MUST:

-   Provide an accurate description of the currently proprosed change, or instructions to recreate it.

> -   Provide the ability to review a change before it is accepted.

Author: reviewing the change before it becomes canon seems pretty critical, as does recording actors, and timestamps. 

I can definitely see moving the approval and approval timestamp / revision record keeping to a higher level. 💯

| Approver | An actor that approves a particular change to the source.
| Merger | An actor that applies a change to the source. This typically involves creating the new revision and updating a branch. This person may be the proposer or a different trusted person, depending on the version control platform.

## Safe Expunging Process

Administrators have the ability to expunge (remove) content from a repository and its change history without leaving a record of the removed content (to accommodate legal or privacy compliance requirements).
This includes changing files, history, or changing references in git.
When used as an attack, this is called “repo hijacking” (or “repo-jacking”) and is one of the primary threats source provenance attestations protect against.

TODO: Determine how organizations can provide transparency around this process.

## Levels

> ### Level 1: Version controlled

Author: >  does "repository" always have to be hosted on an SCP?

probably not. I think the requirement is that the SCP needs to have enough context to write the attestations. 

> a stand-in to mean the "main" or "canonical" copy of a repository

I'm not sure if it helps, but I was thinking something similar to that and have added a section below about branches to clarify things. What do you think?

> ### Level 1: Version controlled

Reviewer 6: > I'm not sure if it helps, but I was thinking something similar to that and have added a section below about branches to clarify things. What do you think?

I think that helps, perhaps it could also say, "for example, the hosted copy on an SCP" or so?

> probably not. I think the requirement is that the SCP needs to have enough context to write the attestations.

I see what you mean, but perhaps we have SCP as a tie-in to the canonical location point on a level by level basis? That way, we can more cleanly support cases where everyone signs over their actions themselves (code review attestations, signed pushes, etc.) and meet requirements _without_ the SCP. Given the impact to usability, this would probably be at a pretty high level, but if someone can meet it, they shouldn't be limited by the fact they don't use an SCP.

I can't mark this unresolved, could you? Thanks!

Summary: The source is stored and managed through a modern version control system with a revision process.

Intended for: Organizations wanting to easily and quickly gain some benefits of SLSA and better integrate with the SLSA ecosystem without changing their source workflows.

Benefits: Version control solves software development challenges ranging from change attribution to effective collaboration.
It is a software development best practice with more benefits than we can discuss here.

Requirements: The combination of SCP and VCS MUST provide:

#### Immutable revisions

There exists a deterministic way to identify a particular revision.

> This is usually a combination of the repository ID and revision ID.

Reviewer: Oh, I had imagined the repository ID is the URL (e.g. `https://github.com/slsa-framework/slsa`).

Having it be something that's meaningful to humans would be pretty beneficial.

> This is usually a combination of the repository ID and revision ID.

Reviewer 6: Yeah, I think that's partly what I was getting at. I wonder if there's an endpoint to translate the numeric ID to the URL or if the current definition for repository (even without the ID) inherently has a URL associated with it because we only consider a repository on the SCP at the moment. This is part of the massaging I wonder if I should take a stab at separately.

> This is usually a combination of the repository ID and revision ID.

Reviewer: IMO we should consider the numeric ID to be an implementation detail and not worry about it?

> This is usually a combination of the repository ID and revision ID.

Author: it's an interesting question -- today we definitely have resurrection and rename attacks (so you need the gh repo numeric id to disambiguate), but I think in combination with the git object id, the url is probably fine. 
IE, it doesn't likely matter much if a hijacked repo (with a different numeric id) has the same url and ships the same revision, we can ship the corresponding attestation. I'd want to think through it some more.

When the revision ID is a number or otherwise not a digest, then the SCP MUST document how the immutability of the revision is established.  See also [Use cases for non-cryptographic, immutable, digests](https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md#use-cases-for-non-cryptographic-immutable-digests).
The SCP MUST guarantee that repository IDs track the complete history of changes that occur to the source while hosted on the platform.

#### Identity Management

There exists an identity management system or some other means of identifying actors.
This system may be a federated authentication system (AAD, Google, Okta, GitHub, etc) or custom (gittuf, gpg-signatures on commits, etc).

#### Revision process

There exists a trusted mechanism for modifying the source pointed-to by a [branch](#definitions).
For each [branch](#definitions), the SCP MUST record and keep the full history of changes conducted on this SCP, with exceptions allowed following the [Safe Expunging Process](#safe-expunging-process).

The revision process MUST:

-   Provide an accurate description of the currently proprosed change, or instructions to recreate it.

> -   Provide the ability to require pre-approval from specific actors before a change proposal is accepted.

Reviewer: Not requiring review before merge SGTM (and as you say, I can see why a SCP might still choose to require the review prior to merge).  However "Record reviews / votes submitted before a change proposal is accepted." still seems to require the action take place before the change is merged?

> -   Provide the ability to require pre-approval from specific actors before a change proposal is accepted.

Reviewer 6: Correct, but I think that's generally in line with recording events in the order that they happen in. Perhaps my phrasing is off. I also reckon reviews after it's accepted is a nice-to-have data point but not meaningful in policy decisions about that merge event? I may be missing some use cases though.

-   Record all actors that contributed to the process, see [Source Roles](#source-roles).
-   Record timestamps of critical activities including process start, process completion, reception of change proposals by the SCP, and reviews.

### Level 2: TODO, but maybe: Generates Source Provenance Attestations

Summary: A consumer can ask the SCP for everything it knows about a specific revision of a repository.

Intended for: Organizations that need to enforce policy on consumed source.

> Benefits: Prevents many classes of accidents and pin-to-sha attacks. Provides information to VSA-generation bots.

Reviewer 6: Could you clarify what a pin-to-sha attack is? I'm parsing "pin-to-sha" as "something downstream pins to a specific revision's SHA", and AIUI this is a mitigation for the mutable nature of non content addressed references such as a tag name, which may move to a different SHA.

Requirements:

#### Source provenance attestation

Results from the revision process are codified into a source provenance attestation document.
Must record who, what, when.

##### Who

Source provenance attestation issuers SHOULD use a single identity definiton.

The strongly authenticated idenenty used to login to the SCP MUST be used for the generation of source provenance attestations.

> This is typically the identity used to push to a git server.

Reviewer 6: ```suggestion
This is typically the identity associated with the actor's pushes to a Git server or the reviews they submit on an SCP.
```

Reviewer: How strongly do you feel about the "associated with the actor's pushes to a Git server" change?

It seems like it could be excessively formal while "used to push" seems to get the same thing across.

But maybe there's a technical nuance I'm not aware of that makes "used to push" ambiguous?

Reviewer 6: I think it could get a little more nuanced, which I was planning to take a stab at addressing separately. On another part, for applicable types of identities, the doc mention things like signatures that could tie into this bit with git signed pushes and so on.

Reviewer 6: True, but I think there's still some tension in the former sentence about the identity used in the provenance attestation, since elsewhere the doc indicates other mechanisms are allowed to. Definitely support punting this one to a separate PR / discussion.

Reviewer 6: For clarity, here's what I mean re identity management: https://github.com/slsa-framework/slsa/blob/09b0c673eaee110d8c4bb06ba04c0c5a6a49a3c6/docs/spec/draft/source-requirements.md?plain=1#L89-L93

I also want to note that I can help massage the gittuf bits there as it uses signatures as the authenticating mechanism, but provides the authenticated metadata layer that can determine what signing keys / identities are trusted, so effectively the policy layer.

Reviewer: Gotcha.  One thing that might help is having some way to talk about 'the system' as a whole vs the individual pieces.  That system needs identity management and it might be that that is provided by the SCP, or with something like GitTuf, or via 'control over an email address'?  Then perhaps the details can go into something like https://slsa.dev/spec/v1.0/verifying-systems for source systems?

Reviewer 6: Quite possibly! That's what I hope to propose, and I think that naturally addresses some of the points I raised elsewhere about the attestations we'd use in this track.

Author: yeah, "used to push" is the typical case for the big cloud SCPs. 

If your identity system is signature based, I guess we'd be trying to say "the identity used by the SCP's primary authentication mechanism?"

Ideally no system would use a _mix_ of identities -- like if one commit wasn't signed, it would fall through to trusting the `committer` or something.

Requirements:
**[Strong authentication]** User accounts that can modify the source or the project's configuration must use multi-factor authentication or its equivalent.

**[Verified timestamps]** Each entry in the change history must contain at least one timestamp that is determined by the source control platform and cannot be modified by clients. It MUST be clear in the change history which timestamps are determined by the source control platform.
#### Source provenance attestation

Results from the revision process are codified into a source provenance attestation document.
Must record who, what, when.

##### Who

Source provenance attestation issuers SHOULD use a single identity definiton.

The strongly authenticated idenenty used to login to the SCP MUST be used for the generation of source provenance attestations.
This is typically the identity used to push to a git server.

Other forms of identity MAY be included as informational.
Examples include a git commit "author" and "committer," and a gpg signature's "user id."
These forms of identity are user-provided and not typically verified by the issuer.

> **[Retained history]** The change history MUST be preserved as long as the source is hosted on the source control system. The source MAY migrate to another source control system, but the organization MUST retain the change history if possible. It MUST NOT be possible for persons to delete or modify the change history, even with multi-party approval, except by trusted platform admins following an established deletion policy.

Author: we could move this down to level 1, or just mention that it was probably provided by a level 1 vcs

> Must record the proposal provided to reviewers for approval or instructions to reproduce it.

Reviewer 6: Should this include things like test run results?

Reviewer: I think test results should be out of scope for the source track?  But that doesn't mean users who care about test results couldn't piggy back on what's being built here.

Zac and I had discussed the other day that the source track is probably best kept to "we know how these revisions were created" (or something like that).

Reviewer 6: That's a fair point, though it does raise the question of where we require automated checks as a gating mechanism. We can table that for a later discussion, though.

> Must record when commits and review are received by the server.

Reviewer 6: ```suggestion
Must record when commits and reviews are received by the server.
```

Do we care about the time an event happened or relative ordering? Is it enough to say "review event X happened before push event Y which happened before merge event Z"? Or is the intent to eventually tie this back to other types of infrastructure / platform logs where we'd need a strong source of absolute time?

> Summary: All changes to the source are approved by two trusted actors prior to submission.

Reviewer 6: ```suggestion
Summary: All changes to the source are approved by two trusted actors prior to acceptance.
```

> All changes to the source are approved by two authorized actors prior to submission.

Reviewer 6: ```suggestion
All changes to the source are approved by two authorized actors prior to acceptance.
```

> User accounts that can perform code reviews MUST use two-factor authentication or its equivalent.

Reviewer 6: Does this disallow that SCP account from issuing access tokens that have the permission to create reviews?

Reviewer: Are you thinking "a robot account (the SCP) cannot use two-factor auth, so it won't be able to meet this requirement"

If so I think we can/should have language that says that _human_ users must use two-factor.  And define 'trusted robots' that don't need to do so (I think this is already an item on our list to resolve).

Perhaps you're also asking if robots from performing the review themselves?

Reviewer 6: See: https://docs.github.com/en/rest/pulls/reviews?apiVersion=2022-11-28#create-a-review-for-a-pull-request as how we could use the token to submit a review.

Reviewer: Interesting point.  So it comes down to "when is the second factor required?"

I seem to recall that right now GitHub doesn't require the second factor at review time but does require it for initial login.  _That_ would seem to be compatible with the PAT approach you outline.

I suspect requiring the second factor at time of review would be too much for Level 1 (and maybe even level 2)?

Reviewer 6: No, sorry I think I wasn't clear enough. My reading is the second factor requirement protects against a password leak enabling an attacker to approve a PR on behalf of the actual approver, likely on a new machine altogether where there isn't an active logged in session. The 2FA requirement would block this, but only if we also explicitly close the door on PATs that can submit reviews. A PAT leak, akin to a password leak, would allow the attacker to authenticate as the approver without 2FA just to submit the review where a session doesn't already exist.

Having thought about it some more, should the spec even require 2FA? Can we leave it more abstract in having strong ways to authenticate, depending on the mechanism used?

This also ties into other discussions in this PR about the model used: can a code review only happen on an SCP? I think there's an implicit (possibly explicit, I need to re-read) assumption that we may not want to make. The spec should likely only require approvals, rather than mandate how the review happens and how reviewers are authenticated? As with the other points, happy to take a stab at it in the follow on PR if that's simpler.
